### ISSUE TITLE
helm disables cua-mode

### ISSUE THREAD

**Issue by krux02** [emacs-helm/helm#2079]
First of all. This is very hard and non-deterministic bug. And it only happens on startup, but not every time. And then, when I want to reproduce it, it doesn't reproduced. I configured my emacs to use ``cua-mode`` all the time, I depend on it. But then when I started emacs and hit ``C-v`` the page scrolled by a page, verry irritating and confusing.  Then when I checked the value for ``cua-mode`` in ``customize`` it said changed outside of ``customize``. So I addad a debugger to see where that value got changed, and it got changed in the helm code.  In many circumstances helm just disables ``cua-mode`` temporarily and then sets it back, but aparently it doesn't do so reliably.  I wish helm wouldn't disable cua-mode continuously. I know there is a conflicting key, namely ``C-z``, but I would rather remap ``C-z`` in the helm keyboard map to something else, for example ``C-h`` or ``M-h``, than a constant disable/enable of cua-mode.

Also ``C-v`` to paste into the helm minibuffer doesn't work, because helm disables ``cua-mode``.

All interfaces from helm that I am actually using are the following. The the bug must be in one of them:

```elisp
(global-set-key (kbd "M-x") 'helm-M-x)
(global-set-key (kbd "s-s") 'helm-grep-do-git-grep)
(global-set-key (kbd "s-b") 'helm-buffers-list)
(global-set-key (kbd "C-x C-f") 'helm-find-files)
;; rarely used:
(global-set-key (kbd "s-S") 'helm-do-grep-ag)  
(global-set-key (kbd "s-l") 'helm-locate) 
```

**Comment by thierryvolpiatto** [comment]
Arne Döring <notifications@github.com> writes:

> First of all. This is very hard and non-deterministic bug. And it only
> happens on startup, but not every time. And then, when I want to
> reproduce it, it doesn't reproduced. I configured my emacs to use
> ``cua-mode`` all the time, I depend on it. But then when I started
> emacs and hit ``C-v`` the page scrolled by a page, verry irritating
> and confusing.  Then when I checked the value for ``cua-mode`` in
> ``customize`` it said changed outside of ``customize``. So I addad a
> debugger to see where that value got changed, and it got changed in
> the helm code.  In many circumstances helm just disables ``cua-mode``
> temporarily and then sets it back, but aparently it doesn't do so
> reliably.

Helm disable cua-mode and then reenable it (if it was enabled) on all
helm sessions, it is guaranteed by an unwind-protect, so I doubt it is
helm that disable cua-mode globally.
From emacs-helm.sh, I can't reproduce:
After starting emacs from emacs-helm.sh, I do M-x cua-mode, then I check
the value of cua-mode (the variable), it is t.
Then I start a helm session (helm-find-files, M-x or whatever), then I
quit this helm session and I eval again cua-mode, it is still t, it
didn't change.

So please try from emacs-helm.sh to see if cua-mode is indeed disabled.

> Also ``C-v`` to paste into the helm minibuffer doesn't work, because
> helm disables ``cua-mode``.

Yes that's true, but at this stage of helm development don't expect such
a change. You can bind C-v to yank in helm-map to workaround this.

--
Thierry

**Comment by krux02** [comment]
As I said, it is non-deterministic. When I start emacs, it might happen that cua-mode is not enabled. It only happens during startup. When I enable cua-mode after that it doesn't disable anymore. But I couldn't reliably reproduce this problem at all. It just comes up again and again. All I could do was to watch whenever something changes the variable ``cua-mode``, and it is indeed helm who is resposible for this. But this indeed no guarantee that helm is resposible for this bug, it could be that something else disables ``cua-mode`` without properly recovering from it, but then if that is the case then the constant changing of ``cua-mode`` within ``helm`` prevents me from finding the actual cause for this weird behavior. 

After all, I don't want that helm changes ``cua-mode``.  It causes problems for me.

**Comment by thierryvolpiatto** [comment]
Arne Döring <notifications@github.com> writes:

> As I said, it is non-deterministic. When I start emacs, it might
> happen that cua-mode is not enabled. It only happens during
> startup. When I enable cua-mode after that it doesn't disable
> anymore. But I couldn't reliably reproduce this problem at all. It
> just comes up again and again. All I could do was to watch whenever
> something changes the variable ``cua-mode``, and it is indeed helm who
> is resposible for this.

It is not.  The code guarantees that helm doesn't disable globally
cua-mode, it only disables it locally while helm is running and as soon
you quit or exit helm it is guaranteed cua-mode is reenabled, as I said
you can easily verify this from emacs-helm.sh by evaling cua-mode after
a helm session. 

So you will have to look somewhere else in your installed packages to
see which one is the culprit, I suggest you bissect your init file until
you find the bug (comment all the file and uncomment progressively until
you hit the bug).

--
Thierry

**Comment by krux02** [comment]
I just had it again today. Cua-mode was disabled after starting ``emacs``. It was only after the first, boot of emacs. Then I closed emacs, and opened it up again, and cua-mode was enabled. I did not change anything in ``init.el``.

**Comment by krux02** [comment]
I found the cause of the bug. Aparently it was not helm or projectile. It was undo-tree that caused the stack overflow when it tried to load the persistent history. Loading the history crashes in the C stack and then helm does not revert properly after that crash.

**Comment by krux02** [comment]
Just for the log, just now ``cua-mode`` has been disabled again without my approval.  By now ``undo-tree`` can't be the cause anymore, because I deleted ``undo-tree`` from my system.

**Comment by thierryvolpiatto** [comment]
Arne Döring <notifications@github.com> writes:

> Just for the log, just now cua-mode has been disabled again without my
> approval.

Hmm, that's bad but hardly see how it could be helm fault, here a sample
code showing basically what helm does:

    ;; Sample code
    (defun how-helm-disable-cua-mode ()
      (let ((old-cua cua-mode))
        (when cua-mode (cua-mode -1))
        (unwind-protect
             (do-something-here)
          (when old-cua
            (cua-mode 1)))))

So I guess you have something else disabling cua-mode.

Did you try reproducing from emacs-helm.sh ?

-- 
Thierry

**Comment by krux02** [comment]
I found a comment in the sources of helm. that shows some attitude towards this issue:

```elisp
    ;; cua-mode override local helm bindings.
    ;; disable this stupid thing if enabled.
```
`helm.el  2364`

I never asked to disable cua-mode temporarily. I even think it is a bad thing, because cua-bindings should be available in the minibuffer. I want to be able to insert with C-v and undo with C-z. I never got used to use C-z for anything else than undo.

Also did I find that before calling (cua-mode 1) a lot of other stuff is done. If any of the other things failes before cua-mode is reverted, cua-mode remains disabled.

`helm.el  2409`
```elisp
      (when (fboundp 'advice-remove)
        (advice-remove 'tramp-read-passwd #'helm--suspend-read-passwd)
        (advice-remove 'ange-ftp-get-passwd #'helm--suspend-read-passwd)
        (advice-remove 'epa-passphrase-callback-function #'helm--suspend-read-passwd)
        (advice-remove 'cua-delete-region #'cua-delete-region--advice)
        (advice-remove 'copy-region-as-kill #'copy-region-as-kill--advice))
      (helm-log "helm-alive-p = %S" (setq helm-alive-p nil))
      (helm--remap-mouse-mode -1)       ; Reenable mouse bindings.
      (setq helm-alive-p nil)
      (setq helm--buffer-in-new-frame-p nil)
      ;; Reset helm-pattern so that lambda's using it
      ;; before running helm will not start with its old value.
      (setq helm-pattern "")
      (setq helm--ignore-errors nil)
```

And as mentioned earlier, "Re-entering top level after C stack overflow" does not execute UNWINDFORMS of `unwind-protect`. It is just that all variables except `cua-mode` are set to their correct value after the next helm invocation.

**Comment by thierryvolpiatto** [comment]
I removed all advices and workarounds for `cua-mode` so don't complain if something is not working around this.

**Comment by npostavs** [comment]
> And as mentioned earlier, "Re-entering top level after C stack overflow" does not execute UNWINDFORMS of unwind-protect

This is [Bug#31995](https://debbugs.gnu.org/cgi/bugreport.cgi?bug=31995), by the way. I don't believe it's practically fixable, but in the `emacs-26` branch (which will be released as 26.2) `read` no longer takes so much stack space, so loading undo-tree history should no longer cause an overflow (unless you have a much bigger history).

**Comment by krux02** [comment]
It seems like this fixed it. C-v in helm works. I did not yet have an accidental cua-mode disable.  Thanks for the patch.  But I still wonder why it was introduced in the first place. Did someone create a bug report about cua-mode and C-z?

**Comment by thierryvolpiatto** [comment]
Arne Döring <notifications@github.com> writes:

> Thanks for the patch. But I still wonder why it was introduced in the
> first place.

It was introduced at the time where the keybindings was not handled, at
this time there was only one keymap `helm-map` and nothing else to
handle minor modes.

> Did someone create a bug report about cua-mode and C-z?

No. Here when I enable cua-mode C-z is undoing, isn't it the case for
you?

-- 
Thierry

**Comment by thierryvolpiatto** [comment]
Noam Postavsky <notifications@github.com> writes:

>     And as mentioned earlier, "Re-entering top level after C stack overflow" does not execute UNWINDFORMS of unwind-protect
>
> This is Bug#31995, by the way. I don't believe it's practically
> fixable, but in the emacs-26 branch (which will be released as 26.2)
> read no longer takes so much stack space, so loading undo-tree history
> should no longer cause an overflow (unless you have a much bigger
> history).

Thanks for the infos.

--
Thierry

**Comment by krux02** [comment]
@thierryvolpiatto yes C-z is undo for me as well. I was just wondering if somebody complained that C-z was not doing helm-execute-persistent-action in a helm buffer when cua-mode was enabled.  And I don't really understand your explanation, but I also don't really deeply understand emacs anyway.
